## Outlier detection & removal using z-score & standard deviation using pandas:-

In [3]:
import pandas as pd
from matplotlib import pyplot as plt

# We are going to use heights dataset from kaggle.com dataset has heights and weights
both but i have removed to make it simple

In [8]:
df=pd.read_csv("heights.csv")

In [9]:
df

Gender     Height
0       Male  73.847017
1       Male  68.781904
2       Male  74.110105
3       Male  71.730978
4       Male  69.881796
...      ...        ...
9995  Female  66.172652
9996  Female  67.067155
9997  Female  63.867992
9998  Female  69.034243
9999  Female  61.944246

[10000 rows x 2 columns]

In [11]:
df.Height.mean()

66.367559754866

In [12]:
df.Height.std()

3.847528120795573

Here the mean is 66.37 and std is 3.84

## (1) Outlier detection and removal using 3 standard deviation

On the ways we can remove outliers is remove any data points that are beyond 3

standard deviation from mean. Which means we can come up with following upper

and lower bounds

In [13]:
upper_limit = df.Height.mean() + 3*df.Height.std()
upper_limit

77.91014411725271

In [14]:
lower_limit = df.Height.mean() - 3*df.Height.std()
lower_limit

54.824975392479274

Here are the outliers that are beyond 3 std dev from mean

In [15]:
df[(df.Height>upper_limit)|(df.Height<lower_limit)]

Gender     Height
994     Male  78.095867
1317    Male  78.462053
2014    Male  78.998742
3285    Male  78.528210
3757    Male  78.621374
6624  Female  54.616858
9285  Female  54.263133

Above the height on higher end is 78 inch which is around 6ft 6 inch. Now that is
quite inusal height. There are people who have this height but it is very uncommaon 
and it is ok if you remove those data points.Similarly on lower end it is 54 inch
which is around 4ft 6 inch. While this is also a legitimate height you don't find
many people having this height so it is safe toconsider both of these cases as outliers.

## Now remove these outliers and generate new dataframe

In [16]:
df_no_outlier_std_dev = df[(df.Height<upper_limit) & (df.Height>lower_limit)]
df_no_outlier_std_dev.head()

Gender     Height
0   Male  73.847017
1   Male  68.781904
2   Male  74.110105
3   Male  71.730978
4   Male  69.881796

In [17]:
df_no_outlier_std_dev.shape

(9993, 2)